leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumObservationRegistryImpl.java

```diff
@@ -64,6 +66,7 @@
 
     private final List<ObservationRegistryListener> listeners = new CopyOnWriteArrayList<>();
     private final Map<KeyToken, Observation> observationsByToken = new ConcurrentHashMap<>();
+    private final Lock observationLock = new ReentrantLock();
```

Comment:
notifying Observation listeners is outside the locks. At the end the same can also be achieved using synchronized blocks. At the end its just a religious thing which one you chose.